s3datarepository/users.go

```diff
@@ -0,0 +1,447 @@
+		if err != nil {
+			return nil, ErrCode("failed to getting access keys for user "+aws.StringValue(u.UserName), err)
+		}
+
+		keys := make(map[string]string, len(keyOut.AccessKeyMetadata))
+		for _, k := range keyOut.AccessKeyMetadata {
+			keys[aws.StringValue(k.AccessKeyId)] = aws.StringValue(k.Status)
+		}
+
+		output[*u.UserName] = struct {
+			Keys map[string]string `json:"keys"`
+		}{keys}
+	}
+
+	log.Debugf("returning map of users for dataset %s: %+v", id, output)
+
+	return output, nil
+}
+
+// listGroupsUsers lists the users that belong to a group
+func (s *S3Repository) listGroupsUsers(ctx context.Context, groupName string) ([]*iam.User, error) {
+	users := []*iam.User{}
+	if groupName == "" {
+		return users, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty group name"))
+	}
+
+	log.Infof("listing iam users for group %s", groupName)
+
+	input := &iam.GetGroupInput{GroupName: aws.String(groupName)}
+
+	truncated := true
+	for truncated {
+		output, err := s.IAM.GetGroupWithContext(ctx, input)
+		if err != nil {
+			return users, ErrCode("failed to getting users for group "+groupName, err)
+		}
+
+		truncated = aws.BoolValue(output.IsTruncated)
+		users = append(users, output.Users...)
+		input.Marker = output.Marker
+	}
+
+	return users, nil
+}
+
+func (s *S3Repository) CreateUser(ctx context.Context, id string) (interface{}, error) {
+	if id == "" {
+		return nil, apierror.New(apierror.ErrBadRequest, "invalid input", errors.New("empty id"))
+	}
+
+	name := id
+	if s.NamePrefix != "" {
+		name = s.NamePrefix + "-" + name
+	}
+
+	log.Infof("creating user of the s3datarepository %s", name)
+
+	policyDoc, err := s.temporaryAccessPolicy(id)
+	if err != nil {
+		return nil, ErrCode("generate temporary access policy for dataset "+id, err)
+	}
+
+	log.Debugf("generated temporary access policy document %s", string(policyDoc))
+
+	// setup rollback function list and defer execution
+	var rollBackTasks []func() error
+	defer func() {
+		if err != nil {
+			log.Errorf("recovering from error creating user in s3datarepository: %s, executing %d rollback tasks", err, len(rollBackTasks))
+			rollBack(&rollBackTasks)
+		}
+	}()
+
+	policyOutput, err := s.IAM.CreatePolicyWithContext(ctx, &iam.CreatePolicyInput{
+		Description:    aws.String("Temporary access policy for dataset " + id),
+		Path:           aws.String(s.IAMPathPrefix),
+		PolicyDocument: aws.String(string(policyDoc)),
+		PolicyName:     aws.String(name + "-DsTmpPlc"),
+	})
+
+	if err != nil {
+		return nil, ErrCode("create temporary access policy for dataset "+id, err)
+	}
+
+	if err := s.IAM.WaitUntilPolicyExistsWithContext(ctx, &iam.GetPolicyInput{PolicyArn: policyOutput.Policy.Arn}); err != nil {
+		return nil, ErrCode("waiting for temporary access policy to exist for dataset "+id, err)
+	}
+
+	// append policy delete to rollback tasks
+	rollBackTasks = append(rollBackTasks, func() error {
```

Comment:
I wonder if this rollback should be added _before_ the Wait, since the Create call actually succeeded and we're just waiting for the policy to propagate at this point.

Reply:
my thinking was if the `WaitUntilPolicyExistsWithContext` fails, the policy shouldn't exist 🤔 so if we fail there we don't want to try to delete the policy?  i guess we don't do anything but  log rollback errors, so maybe its okay.  with that context, what do you think?

Reply:
I thought by the time the rollback executes it might be already available, although not very likely. But it shouldn't hurt to try and delete it anyway (I think the rollback continues even if that step fails).
In this case probably doesn't matter, I just think logically it makes more sense to add the rollback right after the create succeeds.